clippy_lints/src/unicode.rs

```diff
@@ -114,6 +114,12 @@ fn check_str(cx: &LateContext<'_>, span: Span, id: HirId) {
         );
     }
     if string.chars().any(|c| c as u32 > 0x7F) {
+        // Tests store their source in a literal, causing non-ascii comments to be linted and
+        // `#[allow]`s to no longer be respected.
```

Comment:
Ah so it's not the whole test that's getting stringified, what's happening is tests get expanded to something like

```rust
#[rustc_test_marker]
pub const test_name: test::TestDescAndFn =
    test::TestDescAndFn {
        desc: test::TestDesc {
            name: test::StaticTestName("test_name"),
            ...
        },
        testfn: test::StaticTestFn(|| test::assert_test_result(test_name())),
    };

#[allow(clippy::non_ascii_literal)]
fn test_name() {
    // 悲
}
```

It's firing on the `"test_name"` literal, because it has the span of the test function and the lint uses `snippet` to check the source, so it sees the string

```
fn test_name() {
    // 悲
}
```

Maybe this lint should be an early pass, that would give access to the `Lit::token` which isn't (un?)escaped without reading back any snippets

Reply:
Thanks for looking into it! It's a late pass since it makes use of `is_lint_allowed`, is there a way to check that in an early pass?

Reply:
Ooh I didn't realise that required a late context

The first one I'd say could be removed, I find it surprising that it's applying the result of a separate lint in its suggestion

The second allows an early exit, so removing that may have a minor perf hit but I can't imagine it's noticeable. You could replace the `string.chars().zip(...)` with `unicode_normalization::is_nfc` that has a fast path to make up for it

Reply:
I'm not sure if it can be done this way. `snippet` is just reading from the source, but it looks like when using `Lit::token` it's already escaped, so none of the lints actually fire.

Reply:
It should have both, an [`rustc_ast::ast::Lit`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/struct.Lit.html) has  the `token` field of [`rustc_ast::token::Lit`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/token/struct.Lit.html), where `symbol` should be as it appears in source code

That there's two `Lit` structs and two `LitKind`s, and that they contain one another is a bit confusing

If it turns out to be more trouble than it's worth, it'd be reasonable to stick with the current solution, maybe using `clippy_utils::macros::span_is_local` to also catch other macros from elsewhere

Reply:
I don't know if I'm doing something wrong, but `Symbol::as_str` is producing the literal characters "\u{200b}" instead of the ZWSP character.

Reply:
Oh that is weird, for the literal `"zws(​) \x7e"` (has a zws between the parenthesis) it's

```rust
Lit {
    token: Lit {
        kind: Str,
        symbol: "zws(\u{200b}) \\x7e",
        suffix: None,
    },
    kind: Str(
        "zws(\u{200b}) ~",
        Cooked,
    ),
    span: t.rs:4:13: 4:26 (#0),
}
```

I wonder why the tokens are escaped like that. But yeah in that case best to stick with the late pass